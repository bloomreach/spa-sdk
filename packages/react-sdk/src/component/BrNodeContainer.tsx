/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';
import {
  BrContainerBox,
  BrContainerInline,
  BrContainerNoMarkup,
  BrContainerOrderedList,
  BrContainerUnorderedList,
} from '../cms';

export function BrNodeContainer(props: React.PropsWithChildren<BrProps<Container>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    switch (type) {
      case TYPE_CONTAINER_INLINE:
        return BrContainerInline;
      case TYPE_CONTAINER_NO_MARKUP:
        return BrContainerNoMarkup;
      case TYPE_CONTAINER_ORDERED_LIST:
        return BrContainerOrderedList;
      case TYPE_CONTAINER_UNORDERED_LIST:
        return BrContainerUnorderedList;
      default:
        return BrContainerBox;
    }
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
