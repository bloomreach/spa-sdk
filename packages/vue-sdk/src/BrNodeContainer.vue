<!--
  Copyright 2020 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <div v-if="component.getType() === TYPE_CONTAINER_INLINE">
    <slot />
  </div>

  <div v-else-if="component.getType() === TYPE_CONTAINER_NO_MARKUP">
    <slot />
  </div>

  <div v-else-if="component.getType() === TYPE_CONTAINER_ORDERED_LIST">
    <slot />
  </div>

  <div v-else-if="component.getType() === TYPE_CONTAINER_UNORDERED_LIST">
    <slot />
  </div>

  <div v-else>
    <slot />
  </div>
</template>

<script lang="ts">
import {
  Container,
  Page,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { Component, InjectReactive, Inject, Vue } from 'vue-property-decorator';

@Component({
  computed: {
    component(this: BrNodeContainer) {
      return this.component$();
    },
  },
  data: () => ({
    TYPE_CONTAINER_INLINE,
    TYPE_CONTAINER_NO_MARKUP,
    TYPE_CONTAINER_ORDERED_LIST,
    TYPE_CONTAINER_UNORDERED_LIST,
  }),
})
export default class BrNodeContainer extends Vue {
  @Inject() private component$!: () => Container;

  private component!: Container;

  @InjectReactive() private page!: Page;
}
</script>
