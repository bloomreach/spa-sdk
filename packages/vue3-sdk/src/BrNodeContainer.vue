<!--
  - Copyright 2023 Bloomreach
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -   http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <component
    v-if="componentType && componentType in mapping"
    :is="mapping[componentType]"
    :component="component"
    :page="page"
  >
    <slot/>
  </component>

  <br-container-inline v-else-if="componentType === TYPE_CONTAINER_INLINE" :component="component" :page="page">
    <slot/>
  </br-container-inline>

  <br-container-no-markup
    v-else-if="componentType === TYPE_CONTAINER_NO_MARKUP"
    :component="component"
    :page="page"
  >
    <slot/>
  </br-container-no-markup>

  <br-container-ordered-list
    v-else-if="componentType === TYPE_CONTAINER_ORDERED_LIST"
    :component="component"
    :page="page"
  >
    <slot/>
  </br-container-ordered-list>

  <br-container-unordered-list
    v-else-if="componentType === TYPE_CONTAINER_UNORDERED_LIST"
    :component="component"
    :page="page"
  >
    <slot/>
  </br-container-unordered-list>

  <br-container-box v-else :component="component" :page="page">
    <slot/>
  </br-container-box>
</template>

<script setup lang="ts">
import { component$, mapping$, page$ } from '@/providerKeys';
import type { Container } from '@bloomreach/spa-sdk';
import {
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { computed, inject } from 'vue';
import BrContainerBox from './BrContainerBox.vue';
import BrContainerInline from './BrContainerInline.vue';
import BrContainerNoMarkup from './BrContainerNoMarkup.vue';
import BrContainerOrderedList from './BrContainerOrderedList.vue';
import BrContainerUnorderedList from './BrContainerUnorderedList.vue';

const page = inject(page$);
const component = inject<Container>(component$);
const mapping = inject(mapping$, {});
const componentType = computed(() => component?.getType());
</script>
