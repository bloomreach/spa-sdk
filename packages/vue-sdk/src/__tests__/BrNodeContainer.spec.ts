/*
 * Copyright 2020-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Component, Container, Page } from '@bloomreach/spa-sdk';
import {
  TYPE_CONTAINER_BOX,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { shallowMount } from '@vue/test-utils';
import type { GlobalMountOptions } from '@vue/test-utils/dist/types';
import type { Mocked } from 'vitest';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { defineComponent, nextTick, ref } from 'vue';
import type { PropType } from 'vue';
import type { BrMapping } from '../../typings';
import BrNodeContainer from '../BrNodeContainer.vue';
import { component$, mapping$, page$ } from '../providerKeys';

describe('BrNodeContainer', () => {
  let component: Mocked<Container>;
  let mapping: BrMapping;
  let page: Mocked<Page>;
  let global: GlobalMountOptions;

  const SomeContainer = defineComponent({
    name: 'some-container',
    template: `
      <div class="some-container"/>
    `,
    props: {
      page: {
        type: Object as PropType<Page>,
      },
      component: {
        type: Object as PropType<Component>,
      },
    },
  });

  beforeEach(() => {
    component = {
      getType: vi.fn(),
    } as unknown as typeof component;
    mapping = {};
    page = {
      isPreview: vi.fn(),
    } as unknown as typeof page;

    global = {
      provide: {
        [page$ as symbol]: ref(page),
        [mapping$ as symbol]: mapping,
        [component$ as symbol]: ref(component),
      },
    };
  });

  describe('render', () => {
    it('should render a mapped container', async () => {
      component.getType.mockReturnValue('custom' as ReturnType<typeof component.getType>);
      mapping.custom = SomeContainer;

      const wrapper = shallowMount(BrNodeContainer, { global, slots: { default: ['<div />'] } });
      await nextTick();

      const props = wrapper.findComponent(SomeContainer).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toEqual(component);
      expect(props.page).toEqual(page);
    });

    it.each`
      type
      ${TYPE_CONTAINER_BOX}
      ${TYPE_CONTAINER_INLINE}
      ${TYPE_CONTAINER_NO_MARKUP}
      ${TYPE_CONTAINER_ORDERED_LIST}
      ${TYPE_CONTAINER_UNORDERED_LIST}
    `('should render a container based on the type', async ({ type }) => {
      component.getType.mockReturnValue(type);

      const wrapper = shallowMount(BrNodeContainer, { global, slots: { default: ['<div />'] } });
      await nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
