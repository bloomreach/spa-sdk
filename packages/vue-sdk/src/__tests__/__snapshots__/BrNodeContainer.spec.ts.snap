// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BrNodeContainer > render > should render a container based on the type 1`] = `
"<br-meta-stub>
  <br-container-box-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </br-container-box-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainer > render > should render a container based on the type 2`] = `
"<br-meta-stub>
  <br-container-inline-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </br-container-inline-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainer > render > should render a container based on the type 3`] = `
"<br-meta-stub>
  <br-container-no-markup-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </br-container-no-markup-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainer > render > should render a container based on the type 4`] = `
"<br-meta-stub>
  <br-container-ordered-list-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </br-container-ordered-list-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainer > render > should render a container based on the type 5`] = `
"<br-meta-stub>
  <br-container-unordered-list-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </br-container-unordered-list-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainer > render > should render a mapped container 1`] = `
"<br-meta-stub>
  <some-container-stub component=\\"[object Object]\\" page=\\"[object Object]\\">
    <div></div>
  </some-container-stub>
</br-meta-stub>"
`;
